### ISSUE TITLE
net/http: finishRequest() aborts ongoing background reads

### ISSUE THREAD

**Issue by dhananjay92** [golang/go#17559]
[ *continued from Issue #17547* ]

**TL;DR:** [`abortPendingRead()`](https://github.com/golang/go/blob/master/src/net/http/server.go#L1461&q=abortPendingRead) is called from `finishRequest()`, causing it to abort ongoing reads. This causes `deadline_exceeded` errors. `finishRequest()` should wait for reads to finish in normal case.

Consider following sequence of events from [`http.conn.serve()`](https://github.com/golang/go/blob/master/src/net/http/server.go#L1591):

1. 	`startBackgroundRead()` starts [`backgroundRead()` in another goroutine](https://github.com/golang/go/blob/master/src/net/http/server.go#L592&q=cr.backgroundRead()).

2. 	Assume that `backgroundRead()` from above step does not finish until after the next step. (*Simulated by [adding sleep](https://github.com/dhananjay92/go/commit/689a8d466e731e4e7d4c474be7daaa530a62bddc)*)

3. 	`finishRequest()` called, which calls [`abortPendingRead()`](https://github.com/golang/go/blob/master/src/net/http/server.go#L1461&q=abortPendingRead). abortPendingRead() then proceeds to abort ongoing background reads by [setting the read deadline to `aLongTimeAgo`](https://github.com/golang/go/blob/master/src/net/http/server.go#L624&q=SetReadDeadline(aLongTimeAgo)). This causes the ongoing reads to report `deadline_exceeded`.

/cc @bradfitz @dsnet @mdempsky

**Comment by bradfitz** [comment]
I haven't looked into this yet. To clarify, are you saying this is a regression since faf882d1d427e8c8a9a1be00d8ddcab81d1e848e (golang.org/cl/31173)?

**Comment by dhananjay92** [comment]
Regression: Yes. I think so.
I have a CL pending that should fix this. Sending in a bit.

**Comment by dhananjay92** [comment]
From offline discussion with Brad & Joe: It's okay for finishRequest() to abort backgroundRead() every time. The imeplementation of `net.Conn` should respect this.

This is WAI. So closing this.
